### PR TITLE
Minor build system update

### PR DESCRIPTION
**Pull request by echo-Mike**
## Features:
* Make env var G++ name changed to traditional CXX (now it is possible to setup it from shell as G++ is not valid env var name on unix platform but CXX is)
* Make env var GCC name changed to traditional CC
* Make env var PCAPPP_BUILD_FLAGS now hold flags that are passed to compiler globally across whole library (#109)
* MinGw build environment is now better distinguished (important when used with -std flag as MinGw compilers have a lot of problems/bugs/incompatibilities when used with it)

## CI:
Travis: [Build Status](https://travis-ci.org/echo-Mike/PcapPlusPlus)  
Appveyor: [Build status](https://ci.appveyor.com/project/echo-Mike/pcapplusplus)  

## Requested features:
Add support of modern compilers for target platforms in CI (needed for C++11 and above feature testing) -> Add more target compiles with at least C++11 support

### CODE REVIEW

File: Pcap++/header/PcapDevice.h [hunk 1]
```diff
@@ -4,6 +4,10 @@
 #include "RawPacket.h"
 #include "PcapFilter.h"
 #include "PointerVector.h"
+
```
Comment (seladb) [inline Pcap++/header/PcapDevice.h]: Why are these lines necessary?

File: Examples/Tutorials/Tutorial-PcapFiles/Makefile.windows [hunk 2]
```diff
@@ -2,8 +2,8 @@ include ../../../Dist/mk/PcapPlusPlus.mk
 
 # All Target
 all:
-	g++.exe $(PCAPPP_INCLUDES) -c -o main.o main.cpp
-	g++.exe $(PCAPPP_LIBS_DIR) -static-libgcc -static-libstdc++ -o Tutorial-PcapFiles.exe main.o $(PCAPPP_LIBS)
+	$(CXX) $(PCAPPP_BUILD_FLAGS) $(PCAPPP_INCLUDES) -c -o main.o main.cpp
+	$(CXX) $(PCAPPP_LIBS_DIR) -static-libgcc -static-libstdc++ -o Tutorial-PcapFiles.exe main.o $(PCAPPP_LIBS)
```
Comment (seladb) [inline Examples/Tutorials/Tutorial-PcapFiles/Makefile.windows]: Please put back g++.exe instead of $(CXX) to make it simpler to understand

File: Examples/Tutorials/Tutorial-PcapFiles/Makefile.windows [hunk 3]
```diff
@@ -2,8 +2,8 @@ include ../../../Dist/mk/PcapPlusPlus.mk
 
 # All Target
 all:
-	g++.exe $(PCAPPP_INCLUDES) -c -o main.o main.cpp
-	g++.exe $(PCAPPP_LIBS_DIR) -static-libgcc -static-libstdc++ -o Tutorial-PcapFiles.exe main.o $(PCAPPP_LIBS)
+	$(CXX) $(PCAPPP_BUILD_FLAGS) $(PCAPPP_INCLUDES) -c -o main.o main.cpp
```
Comment (seladb) [inline Examples/Tutorials/Tutorial-PcapFiles/Makefile.windows]: Please put back g++.exe instead of $(CXX) to make it simpler to understand

File: Examples/Tutorials/Tutorial-PcapFiles/Makefile.not_windows [hunk 4]
```diff
@@ -2,8 +2,8 @@ include /usr/local/etc/PcapPlusPlus.mk
 
 # All Target
 all:
-	g++ $(PCAPPP_INCLUDES) -c -o main.o main.cpp
-	g++ $(PCAPPP_LIBS_DIR) -static-libstdc++ -o Tutorial-PcapFiles main.o $(PCAPPP_LIBS)
+	$(CXX) $(PCAPPP_BUILD_FLAGS) $(PCAPPP_INCLUDES) -c -o main.o main.cpp
+	$(CXX) $(PCAPPP_LIBS_DIR) -static-libstdc++ -o Tutorial-PcapFiles main.o $(PCAPPP_LIBS)
```
Comment (seladb) [inline Examples/Tutorials/Tutorial-PcapFiles/Makefile.not_windows]: Please put back g++ instead of $(CXX) to make it simpler to understand

File: Examples/Tutorials/Tutorial-PcapFiles/Makefile.not_windows [hunk 5]
```diff
@@ -2,8 +2,8 @@ include /usr/local/etc/PcapPlusPlus.mk
 
 # All Target
 all:
-	g++ $(PCAPPP_INCLUDES) -c -o main.o main.cpp
-	g++ $(PCAPPP_LIBS_DIR) -static-libstdc++ -o Tutorial-PcapFiles main.o $(PCAPPP_LIBS)
+	$(CXX) $(PCAPPP_BUILD_FLAGS) $(PCAPPP_INCLUDES) -c -o main.o main.cpp
```
Comment (seladb) [inline Examples/Tutorials/Tutorial-PcapFiles/Makefile.not_windows]: Please put back g++ instead of $(CXX) to make it simpler to understand

File: Examples/Tutorials/Tutorial-PacketParsing/Makefile.windows [hunk 6]
```diff
@@ -2,8 +2,8 @@ include ../../../Dist/mk/PcapPlusPlus.mk
 
 # All Target
 all:
-	g++.exe $(PCAPPP_INCLUDES) -c -o main.o main.cpp
-	g++.exe $(PCAPPP_LIBS_DIR) -static-libgcc -static-libstdc++ -o Tutorial-PacketParsing.exe main.o $(PCAPPP_LIBS)
+	$(CXX) $(PCAPPP_BUILD_FLAGS) $(PCAPPP_INCLUDES) -c -o main.o main.cpp
+	$(CXX) $(PCAPPP_LIBS_DIR) -static-libgcc -static-libstdc++ -o Tutorial-PacketParsing.exe main.o $(PCAPPP_LIBS)
```
Comment (seladb) [inline Examples/Tutorials/Tutorial-PacketParsing/Makefile.windows]: Please put back g++.exe instead of $(CXX) to make it simpler to understand

File: Examples/Tutorials/Tutorial-PacketParsing/Makefile.windows [hunk 7]
```diff
@@ -2,8 +2,8 @@ include ../../../Dist/mk/PcapPlusPlus.mk
 
 # All Target
 all:
-	g++.exe $(PCAPPP_INCLUDES) -c -o main.o main.cpp
-	g++.exe $(PCAPPP_LIBS_DIR) -static-libgcc -static-libstdc++ -o Tutorial-PacketParsing.exe main.o $(PCAPPP_LIBS)
+	$(CXX) $(PCAPPP_BUILD_FLAGS) $(PCAPPP_INCLUDES) -c -o main.o main.cpp
```
Comment (seladb) [inline Examples/Tutorials/Tutorial-PacketParsing/Makefile.windows]: Please put back g++.exe instead of $(CXX) to make it simpler to understand

File: Examples/Tutorials/Tutorial-PacketParsing/Makefile.not_windows [hunk 8]
```diff
@@ -2,8 +2,8 @@ include /usr/local/etc/PcapPlusPlus.mk
 
 # All Target
 all:
-	g++ $(PCAPPP_INCLUDES) -c -o main.o main.cpp
-	g++ $(PCAPPP_LIBS_DIR) -static-libstdc++ -o Tutorial-PacketParsing main.o $(PCAPPP_LIBS)
+	$(CXX) $(PCAPPP_BUILD_FLAGS) $(PCAPPP_INCLUDES) -c -o main.o main.cpp
+	$(CXX) $(PCAPPP_LIBS_DIR) -static-libstdc++ -o Tutorial-PacketParsing main.o $(PCAPPP_LIBS)
```
Comment (seladb) [inline Examples/Tutorials/Tutorial-PacketParsing/Makefile.not_windows]: Please put back g++ instead of $(CXX) to make it simpler to understand

File: Examples/Tutorials/Tutorial-PacketParsing/Makefile.not_windows [hunk 9]
```diff
@@ -2,8 +2,8 @@ include /usr/local/etc/PcapPlusPlus.mk
 
 # All Target
 all:
-	g++ $(PCAPPP_INCLUDES) -c -o main.o main.cpp
-	g++ $(PCAPPP_LIBS_DIR) -static-libstdc++ -o Tutorial-PacketParsing main.o $(PCAPPP_LIBS)
+	$(CXX) $(PCAPPP_BUILD_FLAGS) $(PCAPPP_INCLUDES) -c -o main.o main.cpp
```
Comment (seladb) [inline Examples/Tutorials/Tutorial-PacketParsing/Makefile.not_windows]: Please put back g++ instead of $(CXX) to make it simpler to understand

File: Examples/Tutorials/Tutorial-PacketCraftAndEdit/Makefile.windows [hunk 10]
```diff
@@ -2,8 +2,8 @@ include ../../../Dist/mk/PcapPlusPlus.mk
 
 # All Target
 all:
-	g++.exe $(PCAPPP_INCLUDES) -c -o main.o main.cpp
-	g++.exe $(PCAPPP_LIBS_DIR) -static-libgcc -static-libstdc++ -o Tutorial-PacketCraftAndEdit.exe main.o $(PCAPPP_LIBS)
+	$(CXX) $(PCAPPP_BUILD_FLAGS) $(PCAPPP_INCLUDES) -c -o main.o main.cpp
+	$(CXX) $(PCAPPP_LIBS_DIR) -static-libgcc -static-libstdc++ -o Tutorial-PacketCraftAndEdit.exe main.o $(PCAPPP_LIBS)
```
Comment (seladb) [inline Examples/Tutorials/Tutorial-PacketCraftAndEdit/Makefile.windows]: Please put back g++.exe instead of $(CXX) to make it simpler to understand

File: Examples/Tutorials/Tutorial-PacketCraftAndEdit/Makefile.windows [hunk 11]
```diff
@@ -2,8 +2,8 @@ include ../../../Dist/mk/PcapPlusPlus.mk
 
 # All Target
 all:
-	g++.exe $(PCAPPP_INCLUDES) -c -o main.o main.cpp
-	g++.exe $(PCAPPP_LIBS_DIR) -static-libgcc -static-libstdc++ -o Tutorial-PacketCraftAndEdit.exe main.o $(PCAPPP_LIBS)
+	$(CXX) $(PCAPPP_BUILD_FLAGS) $(PCAPPP_INCLUDES) -c -o main.o main.cpp
```
Comment (seladb) [inline Examples/Tutorials/Tutorial-PacketCraftAndEdit/Makefile.windows]: Please put back g++.exe instead of $(CXX) to make it simpler to understand

File: Examples/Tutorials/Tutorial-PacketCraftAndEdit/Makefile.not_windows [hunk 12]
```diff
@@ -2,8 +2,8 @@ include /usr/local/etc/PcapPlusPlus.mk
 
 # All Target
 all:
-	g++ $(PCAPPP_INCLUDES) -c -o main.o main.cpp
-	g++ $(PCAPPP_LIBS_DIR) -static-libstdc++ -o Tutorial-PacketCraftAndEdit main.o $(PCAPPP_LIBS)
+	$(CXX) $(PCAPPP_BUILD_FLAGS) $(PCAPPP_INCLUDES) -c -o main.o main.cpp
+	$(CXX) $(PCAPPP_LIBS_DIR) -static-libstdc++ -o Tutorial-PacketCraftAndEdit main.o $(PCAPPP_LIBS)
```
Comment (seladb) [inline Examples/Tutorials/Tutorial-PacketCraftAndEdit/Makefile.not_windows]: Please put back g++ instead of $(CXX) to make it simpler to understand

File: Examples/Tutorials/Tutorial-PacketCraftAndEdit/Makefile.not_windows [hunk 13]
```diff
@@ -2,8 +2,8 @@ include /usr/local/etc/PcapPlusPlus.mk
 
 # All Target
 all:
-	g++ $(PCAPPP_INCLUDES) -c -o main.o main.cpp
-	g++ $(PCAPPP_LIBS_DIR) -static-libstdc++ -o Tutorial-PacketCraftAndEdit main.o $(PCAPPP_LIBS)
+	$(CXX) $(PCAPPP_BUILD_FLAGS) $(PCAPPP_INCLUDES) -c -o main.o main.cpp
```
Comment (seladb) [inline Examples/Tutorials/Tutorial-PacketCraftAndEdit/Makefile.not_windows]: Please put back g++ instead of $(CXX) to make it simpler to understand

File: Examples/Tutorials/Tutorial-LiveTraffic/Makefile.windows [hunk 14]
```diff
@@ -2,8 +2,8 @@ include ../../../Dist/mk/PcapPlusPlus.mk
 
 # All Target
 all:
-	g++.exe $(PCAPPP_INCLUDES) -c -o main.o main.cpp
-	g++.exe $(PCAPPP_LIBS_DIR) -static-libgcc -static-libstdc++ -o Tutorial-LiveTraffic.exe main.o $(PCAPPP_LIBS)
+	$(CXX) $(PCAPPP_BUILD_FLAGS) $(PCAPPP_INCLUDES) -c -o main.o main.cpp
+	$(CXX) $(PCAPPP_LIBS_DIR) -static-libgcc -static-libstdc++ -o Tutorial-LiveTraffic.exe main.o $(PCAPPP_LIBS)
```
Comment (seladb) [inline Examples/Tutorials/Tutorial-LiveTraffic/Makefile.windows]: Please put back g++ instead of $(CXX) to make it simpler to understand

File: Examples/Tutorials/Tutorial-LiveTraffic/Makefile.windows [hunk 15]
```diff
@@ -2,8 +2,8 @@ include ../../../Dist/mk/PcapPlusPlus.mk
 
 # All Target
 all:
-	g++.exe $(PCAPPP_INCLUDES) -c -o main.o main.cpp
-	g++.exe $(PCAPPP_LIBS_DIR) -static-libgcc -static-libstdc++ -o Tutorial-LiveTraffic.exe main.o $(PCAPPP_LIBS)
+	$(CXX) $(PCAPPP_BUILD_FLAGS) $(PCAPPP_INCLUDES) -c -o main.o main.cpp
```
Comment (seladb) [inline Examples/Tutorials/Tutorial-LiveTraffic/Makefile.windows]: Please put back g++.exe instead of $(CXX) to make it simpler to understand

File: Examples/Tutorials/Tutorial-LiveTraffic/Makefile.non_windows [hunk 16]
```diff
@@ -2,8 +2,8 @@ include /usr/local/etc/PcapPlusPlus.mk
 
 # All Target
 all:
-	g++ $(PCAPPP_INCLUDES) -c -o main.o main.cpp
-	g++ $(PCAPPP_LIBS_DIR) -static-libstdc++ -o Tutorial-LiveTraffic main.o $(PCAPPP_LIBS)
+	$(CXX) $(PCAPPP_BUILD_FLAGS) $(PCAPPP_INCLUDES) -c -o main.o main.cpp
+	$(CXX) $(PCAPPP_LIBS_DIR) -static-libstdc++ -o Tutorial-LiveTraffic main.o $(PCAPPP_LIBS)
```
Comment (seladb) [inline Examples/Tutorials/Tutorial-LiveTraffic/Makefile.non_windows]: Please put back g++ instead of $(CXX) to make it simpler to understand

File: Examples/Tutorials/Tutorial-LiveTraffic/Makefile.non_windows [hunk 17]
```diff
@@ -2,8 +2,8 @@ include /usr/local/etc/PcapPlusPlus.mk
 
 # All Target
 all:
-	g++ $(PCAPPP_INCLUDES) -c -o main.o main.cpp
-	g++ $(PCAPPP_LIBS_DIR) -static-libstdc++ -o Tutorial-LiveTraffic main.o $(PCAPPP_LIBS)
+	$(CXX) $(PCAPPP_BUILD_FLAGS) $(PCAPPP_INCLUDES) -c -o main.o main.cpp
```
Comment (seladb) [inline Examples/Tutorials/Tutorial-LiveTraffic/Makefile.non_windows]: Please put back g++ instead of $(CXX) to make it simpler to understand

File: Pcap++/src/PcapLiveDevice.cpp [hunk 18]
```diff
@@ -1,5 +1,6 @@
 #define LOG_MODULE PcapLogModuleLiveDevice
 
+#include "IpUtils.h"
```
Comment (seladb) [inline Pcap++/src/PcapLiveDevice.cpp]: Why adding this include?

File: mk/PcapPlusPlus.mk.common [hunk 19]
```diff
@@ -15,3 +15,6 @@ PCAPPP_POST_BUILD :=
 # build flags
 PCAPPP_BUILD_FLAGS :=
 
+ifdef PCAPPP_ENABLE_CPP_FEATURE_DETECTION
```
Comment (seladb) [inline mk/PcapPlusPlus.mk.common]: I don't see anywhere in the code where you use the `-DPCAPPP_CPP_FEATURE_DETECTION` flag

File: Examples/Tutorials/Tutorial-PcapFiles/Makefile.windows (hunk 2 above)
Comment (echo-Mike) [reply Examples/Tutorials/Tutorial-PcapFiles/Makefile.windows]: Ther are strong reassons to not to do this:  
1. We have at least 40 publicly available C++ compilers besides g++
2. As i heve mentioned in PR description `CXX` is a **traditionl** name for c++ compiler. My arguments are:  
  * `G++` env var name is not portable across unix systems but `CXX` is (this is my actual motivation as i have had this case on my work): https://stackoverflow.com/questions/2821043/allowed-characters-in-linux-environment-variable-names
*  To enshure You that it is a **traditional** use case i have found next code bases that refers to c++ compiler as `CXX`:
https://github.com/nlohmann/json/blob/develop/Makefile
https://github.com/BVLC/caffe/blob/master/Makefile
https://github.com/grpc/grpc
https://github.com/Microsoft/CNTK/blob/master/Makefile
https://github.com/bitcoin/bitcoin/blob/master/configure.ac  
* In fact this names (`CC and CXX`) are so common that CI (Travis)  is atomatically exports them to test env(please refere to passed tests output):  

**662.9** :
```
50 $ export CXX=clang++
51 $ export CC=clang
```
**662.1** :
```
446 $ export CXX=g++
447 $ export CC=gcc
```

File: Pcap++/src/PcapLiveDevice.cpp (hunk 18 above)
Comment (echo-Mike) [reply Pcap++/src/PcapLiveDevice.cpp]: @seladb Fighting with `-std` bugs on MinGw env

File: mk/PcapPlusPlus.mk.common (hunk 19 above)
Comment (echo-Mike) [reply mk/PcapPlusPlus.mk.common]: @seladb This is the reserve of macro definition for future update - automatic c++11 feature recognition and support. I will consider **build up on** strategy for them where features are integrated and turned on only where they are applicable and they are not substituted or reimplemented in older version (uglier code but no pain with backwards compatibility).

File: Examples/Tutorials/Tutorial-PcapFiles/Makefile.windows (hunk 2 above)
Comment (echo-Mike) [reply Examples/Tutorials/Tutorial-PcapFiles/Makefile.windows]: @seladb But generally i not strongly against that examples may be written as such but i really like consistency too.

File: Pcap++/header/PcapDevice.h (hunk 1 above)
Comment (seladb) [reply Pcap++/header/PcapDevice.h]: Got it. To make it more clear I'd add a comment on above these lines explaining what you just explained here

File: Examples/Tutorials/Tutorial-PcapFiles/Makefile.windows (hunk 2 above)
Comment (seladb) [reply Examples/Tutorials/Tutorial-PcapFiles/Makefile.windows]: I totally understand, that's why I support your idea of replacing `G++` with `CXX` throughout the code. However, in the tutorial examples I intentionally put `g++` for Linux/Mac and `g++.exe` for Windows so it'll be  it easier for people with poor C++ experience to run them.

That's why I asked to put back `g++` and `g++.exe` just for the tutorial examples (and not for all examples)
